Adaptation of H2 network options as per paper

Eddy-JV opened the pull request:
# Closes # (if applicable).

## Changes proposed in this Pull Request
The user now has now the option to choose the following in the `config.yaml` file:

1. **Greenfield**:
   - Only electrical transmission lines as routing for new H2 pipelines.
   - No repurposing.
   - By setting the following:
     - `["sector"]["hydrogen"]["network_routes"]`: "greenfield"
     - `["sector"]["hydrogen"]["gas_network_repurposing"]`: false

2. **Only New H2 pipelines following gas network**:
   - Gas network as a routing for new H2 pipelines.
   - No repurposing.
   - By setting the following:
     - `["sector"]["hydrogen"]["network_routes"]`: "gas"
     - `["sector"]["hydrogen"]["gas_network_repurposing"]`: false

3. **New H2 pipelines following gas network + Repurposing**:
   - Gas network as a routing for new H2 pipelines.
   - Repurposing of existing gas network.
   - By setting the following:
     - `["sector"]["hydrogen"]["network_routes"]`: "gas"
     - `["sector"]["hydrogen"]["gas_network_repurposing"]`: true

4. **New H2 pipelines following electrical transmission lines + Repurposing**:
   - Electrical transmission lines as a proxy for new H2 pipelines.
   - Repurposing of existing gas network.
   - By setting the following:
     - `["sector"]["hydrogen"]["network_routes"]`: "greenfield"
     - `["sector"]["hydrogen"]["gas_network_repurposing"]`: true

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Newly introduced dependencies are added to `envs/environment.yaml` and `envs/environment.docs.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`, `config.tutorial.yaml`, and `test/config.test1.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv` and line references are adjusted in `doc/configuration.rst` and `doc/tutorial.rst`.
- [x] A note for the release notes `doc/release_notes.rst` is amended in the format of previous release notes, including reference to the requested PR.

davide-f commented on scripts/prepare_sector_network.py:
```diff
@@ -329,13 +330,21 @@ def add_hydrogen(n, costs):
 
-        if snakemake.config["clustering_options"]["alternative_clustering"]:
+        # Read the config.pypsa-earth.yaml file
+        yaml = ruamel.yaml.YAML()
+        file_path = "./config.pypsa-earth.yaml"
+        with open(file_path, "r") as file:
+            config_pypsa_earth = yaml.load(file)
```
This can lead to problems unfortunately.
As an alternative, we could place an additional `configfile: "config.pypsa-earth.yaml"` at the beginning of the snakefile to merge the two yaml files before this point.
What do you think?
If we don't do that now, we'll need to do that before the merge

Eddy-JV replied:
I integrated the needed line from config.pypsa-earth.yaml into config.default.yaml in order to avoid using the above method.

Of course this can be nicely solved once we merge both models.

davide-f commented on scripts/prepare_sector_network.py:
```diff
@@ -361,44 +370,99 @@ def add_hydrogen(n, costs):
-                lifetime=costs.at["H2 (g) pipeline repurposed", "lifetime"],
-            )
+            if (
+                not snakemake.config["sector"]["hydrogen"]["network_routes"]
+                == "greenfield"
```
not sure if this leads to the intended behavior.
May it be better
```suggestion
                snakemake.config["sector"]["hydrogen"]["network_routes"]
                != "greenfield"

```
?

Eddy-JV replied:
Done

davide-f commented on scripts/prepare_sector_network.py:
```diff
@@ -361,44 +370,99 @@ def add_hydrogen(n, costs):
+            ):
+                # If NOT greenfield, then the existing pipelines routing is used for new and repurposed piepelines.
+                # Else, existing pipelines routing is used for new piepelines only.
+                n.madd(
```
From a styling point of view, here we have a lot of nested if/elif that leads to a superlong function.
It seems that it may be nicer to define auxiliary functions to take care of the cases and make the code cleaner, e.g.
```suggestion
if (cond):
  do_this_1(...)
elif (cond 2):
  do_this_2(...)
...
```

What do you think?

Eddy-JV replied:
Done

davide-f reviewed (commented):
Great @Eddy-JV :D
I don't go into the functionality as I know you discussed with hazem and tested.

I dropped few comments for possibly improving the code, but I'd be eager to hear your feedback.

Eddy-JV commented:
> Great @Eddy-JV :D I don't go into the functionality as I know you discussed with hazem and tested.
> 
> I dropped few comments for possibly improving the code, but I'd be eager to hear your feedback.

Hi @davide-f , sorry for the late reply. I have updated the PR including your very helpful comments. Thanks for taking the time to check it. 

I guess now, the problem faced by @doneachh is solved. The problem was that some countries do not have any existing gas infra and this resulted in empty dataframes as well as its corresponding error.

But now the solution is: For countries/bus_regions that do not have existing gas infra, New H2 pipelines following electrical transmission lines is used. 

Please check the solution if you have some time and give me a feedback. This way we can already merge the PR.

davide-f commented on scripts/prepare_sector_network.py:
```diff
@@ -361,71 +432,88 @@ def add_hydrogen(n, costs):
-                h2_links.at[name, "bus0"] = buses[0]
-                h2_links.at[name, "bus1"] = buses[1]
-                h2_links.at[name, "length"] = candidates.at[candidate, "length"]
+        if len(h2_links) > 0:
```
This if-case seems quite complex; there should be a way to simplify it.
For the first if (len > 0), it seems we can create a table like this:
| repurposing | greenfield|output|
|--|--|--|
|0|0|add_links_new_H2_pipelines|
|0|1|add_links_elec_routing_new_H2_pipelines|
|1|0|add_links_repurposed_H2_pipelines + add_links_new_H2_pipelines|
|1|1|add_links_repurposed_H2_pipelines + add_links_elec_routing_new_H2_pipelines|


Based on that, in pseudo code, we could do:
``` python
if h2_links and repurposing:
   add_links_repurposed_H2_pipelines()
if greenfield or nor h2_links:
   add_links_elec_routing_new_H2_pipelines()
else:
   add_links_new_H2_pipelines()
```
To my understanding when no h2links are available, only add_links_elec_routing_new_H2_pipelines applies.
Is that correct?

A cleaner code may also be:
``` python
if h2_links:
   if repurposing:
      add_links_repurposed_H2_pipelines()
   if greenfield:
      add_links_elec_routing_new_H2_pipelines()
   else:
      add_links_new_H2_pipelines()
else:
   add_links_new_H2_pipelines()
```

What do you think?

hazemakhalek replied:
Hey Davide, I think the suggestion make total sense. I implemented it already. Basically all the functionality is the same with much less cases to handle.

davide-f reviewed (commented):
Nice :D Added a little comment

hazemakhalek commented:
@davide-f I'd like to merge this, please let me know if you have any final comments
